### PR TITLE
GUACAMOLE-1525: Support for true fullscreen mode and keyboard lock

### PR DESCRIPTION
**Pull request by sirux88**
associated JIRA issue: https://issues.apache.org/jira/browse/GUACAMOLE-1525

see also: GUACAMOLE-989 and GUACAMOLE-124

To start fullscreen mode:
Open gucamole's special menu via Ctrl+Alt+Shift and click fullsceen:
image
("Vollbild" is german for Fullscreen)

### CODE REVIEW

**Comment by necouchman** [comment]
@sirux88 I know you've been frustrated by difficulty of getting PRs merged. Are you still willing to work on this one? I have a couple of minor style issues and then I think this one could be ready to go and close out several long-standing issues.

**Comment by sirux88** [comment]
@necouchman:
If there's a chance to get it merged by time I'm willing to work on this for sure.
Just let me know.

Since it is only on the jira issue but affecs functionality:
1) `True fullscreen mode` should work on each maintained browser as stated here => https://developer.mozilla.org/en-US/docs/Web/API/Fullscreen_API
2) `Keyboard lock` will only work on chromium based browsers => https://developer.mozilla.org/en-US/docs/Web/API/Keyboard/lock

File: guacamole/src/main/frontend/src/translations/en.json [hunk 1]
```diff
@@ -62,6 +62,7 @@
         "ACTION_CANCEL"                    : "@:APP.ACTION_CANCEL",
         "ACTION_CLEAR_CLIENT_MESSAGES"     : "Clear",
         "ACTION_CLEAR_COMPLETED_TRANSFERS" : "Clear",
+        "ACTION_FULLSCREEN"                : "Fullscreen",
```
Comment (necouchman) [inline guacamole/src/main/frontend/src/translations/en.json:65]: I think this should be down below `ACTION_DISCONNECT` by Alphabetical order?

File: guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+/**
+ * A service for providing true fullscreen and keyboard lock support.
+ * Keyboard lock is currently only supported by Chromium based browsers
+ * (Edge >= V79, Chrome >= V68 and Opera >= V55)
+ */
+angular.module('client').factory('guacFullscreen', ['$injector',
+
+    
+    function guacFullscreen($injector) {
```
Comment (necouchman) [inline guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js:28]: Looking at other services, from a style perspective I think you can remove the space between the `angular...` line and the `function...` line, and then indent the function line one more. For example, see:

https://github.com/apache/guacamole-client/blob/d1faaa9605c5eef668da8bf84279d7a88cad5af7/guacamole/src/main/frontend/src/app/client/services/guacClientManager.js#L20-L25

File: guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+        service.isInFullscreenMode=function isInFullscreenMode(){
+            return document.fullscreenElement;
+        }
+
+        // set fullscreen mode
+        service.setFullscreenMode = function setFullscreenMode(state) {
+            if(document.fullscreenEnabled){
+                
+                if(state && !service.isInFullscreenMode()) document.documentElement.requestFullscreen().then(navigator.keyboard.lock()); 
+                else if(!state && service.isInFullscreenMode()) document.exitFullscreen().then(navigator.keyboard.unlock());
```
Comment (necouchman) [inline guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js:51]: These lines are long enough that it probably makes sense to not try to do one-liners, here, and, instead, put the `document..` part on its own line.

Also, please put spaces between the `if` and the opening parenthesis.

File: guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+        var service = {};
+
+        // toggles current fullscreen mode (off if on, on if off)
+        service.toggleFullscreenMode = function toggleFullscreenMode(){
+            if(!service.isInFullscreenMode()){
+                service.setFullscreenMode(true);
+            }else{
+                service.setFullscreenMode(false);
+            }
+        }
```
Comment (necouchman) [inline guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js:39]: * Please put spaces between parenthesis and brackets.
* Please put spaces between the `if` and its opening parenthesis, and between the `else` and its bracket.
* Please don't cuddle the `else` clause with the closing bracket above it.

File: guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+        // check is browser in true fullscreen mode
+        service.isInFullscreenMode=function isInFullscreenMode(){
```
Comment (necouchman) [inline guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js:42]: Spaces around the equals sign, and between `()` and `{`

File: guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+            return document.fullscreenElement;
+        }
+
+        // set fullscreen mode
+        service.setFullscreenMode = function setFullscreenMode(state) {
+            if(document.fullscreenEnabled){
```
Comment (necouchman) [inline guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js:48]: Spaces between `if` and `(` and between `)` and `{`

**Review by necouchman — CHANGES_REQUESTED** [review]
Thanks, @sirux88 - Mostly some style change requests. It can seem like some of these are trivial or pedantic, but when you scale it up to the amount of code, here, it makes a difference after a while for the overall readability of the code, so we tend to be very particular about the style being consisten.

Other than that, unless @mike-jumper or @jmuehlner have any concerns, I think it's good.

File: guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js (hunk 2 above)
Comment (sirux88) [reply guacamole/src/main/frontend/src/app/client/services/guacFullscreen.js]: It's not always done the same coding style. I tried to align it to your example

**Comment by sirux88** [comment]
> Thanks, @sirux88 - Mostly some style change requests. It can seem like some of these are trivial or pedantic, but when you scale it up to the amount of code, here, it makes a difference after a while for the overall readability of the code, so we tend to be very particular about the style being consisten.

I understand that and it's ok. I hope it's fne now @necouchman 

To be honest maybe you should add a linter or [codestyle checker](https://maven.apache.org/plugins/maven-checkstyle-plugin/usage.html) to your project and setup your rules. So you haven't to explain it to contributers (which is annoying I know)

**Comment by necouchman** [comment]
> To be honest maybe you should add a linter or [codestyle checker](https://maven.apache.org/plugins/maven-checkstyle-plugin/usage.html) to your project and setup your rules.

Ah, nice, I will have a look.

**Review by necouchman — APPROVED** [review]
Looks okay to me, now. I will give @mike-jumper and @jmuehlner a while longer to respond with any issues they have, otherwise I'll go ahead and merge it.

**Comment by sirux88** [comment]
Is the failing docker build a problem @necouchman? 
The problem within the build seems not related to this PR.

**Comment by sirux88** [comment]
I fell over an problem concerning browsers not implementing `keyboard.lock` api (e.g. firefox):
If you set fullscreen mode and use the browsers back functionality (e.g. mouse xbutton1) the screen will stay in fullscreen mode.
This might be confusing to some users.

My approach to solve it would be to unload fullscreen mode if the client is "minimized". Meaning guacamole application is going to its homescreen.

But I haven't found an event to register to.
Is there an event like this @necouchman ?

**Comment by sirux88** [comment]
> I fell over an problem concerning browsers not implementing `keyboard.lock` api (e.g. firefox): If you set fullscreen mode and use the browsers back functionality (e.g. mouse xbutton1) the screen will stay in fullscreen mode. This might be confusing to some users.
> 
> My approach to solve it would be to unload fullscreen mode if the client is "minimized". Meaning guacamole application is going to its homescreen.
> 
> But I haven't found an event to register to. Is there an event like this @necouchman ?

I ended up in adding it to `$scope.$on('$destroy', function clientViewDestroyed()` function. 
=> solved in my eyes

**Comment by necouchman** [comment]
> I ended up in adding it to $scope.$on('$destroy', function clientViewDestroyed() function.
=> solved in my eyes

Seems reasonable to me.

File: guacamole/src/main/frontend/src/app/client/controllers/clientController.js [hunk 7]
```diff
@@ -44,7 +44,7 @@ angular.module('client').controller('clientController', ['$scope', '$routeParams
     const requestService         = $injector.get('requestService');
     const tunnelService          = $injector.get('tunnelService');
     const userPageService        = $injector.get('userPageService');
-
+    const guacFullscreen         = $injector.get('guacFullscreen');
```
Comment (necouchman) [inline guacamole/src/main/frontend/src/app/client/controllers/clientController.js:47]: Nitpick, but could you put this service in alphabetical order with the ones above it, and then keep the space below this before the next code comment?

**Review by necouchman — CHANGES_REQUESTED** [review]
One more style change request.

File: guacamole/src/main/frontend/src/app/client/controllers/clientController.js (hunk 7 above)
Comment (mike-jumper) [reply guacamole/src/main/frontend/src/app/client/controllers/clientController.js:47]: This is fairly important, IMHO. Maintaining alphabetical order for things like `#include`, `import`, blocks of retrievals from the AngularJS `$injector`, etc. helps avoid merge conflicts.

Comment (sirux88) [reply guacamole/src/main/frontend/src/app/client/controllers/clientController.js:47]: done

**Review by necouchman — APPROVED** [review]

**Pull request merged**